prio/client.c

```diff
@@ -336,6 +336,64 @@ PrioClient_encode(const_PrioConfig cfg, const bool* data_in,
   return rv;
 }
 
+// Encode long as bools and write them to an array with offset.
+SECStatus
+long_to_bool(bool* dst, long x, int prec, int entry)
+{
+  SECStatus rv = SECSuccess;
+
+  // maximum for b-bit encoding (b = precision)
+  long max = (1l << prec) - 1; // TODO: Compute maximum in PrioConfig?
+
+  // position at which encoding of current entry starts in dst
+  int offset = entry * prec;
+
+  P_CHECKCB(max >= x); // Check wether x is can be encoded with given precision
+  P_CHECKCB(0 <= x);   // Check wether x is non-negative
+
+  // The following block sets b_0, ..., b_(b-1)
+  // set b_0 (the first bit)
+  dst[offset + prec - 1] = (x & 1);
+  // start with prec-2 to shift correctly
+  for (int bit = prec - 2; bit >= 0; bit--) {
+    // shift by one for subsequent bits (b_1, ..., b_(b-1))
+    (x >>= 1);
+    dst[offset + bit] = (x & 1);
+    // from C99 on ">>" produces leading 0s since only unsigned ints are handled
+    // (see ISO/IEC 9899:1999, 6.5.7)
+  }
+
+cleanup:
+  return rv;
+}
+
+SECStatus
+PrioClient_encode_int(const_PrioConfig cfg, const int prec,
+                      const long* data_int, unsigned char** forServerA,
+                      unsigned int* aLen, unsigned char** forServerB,
+                      unsigned int* bLen)
+{
+  SECStatus rv = SECSuccess;
+  bool* data_bool = NULL;
+
+  int num_ints = PrioConfig_numIntEntries(cfg, prec);
+  P_CHECKCB(cfg->num_data_fields == prec * num_ints);
```

Comment:
I was thinking that there might be an integer-overflow issue here, but on second look, it's probably fine.

Reply:
As long as PrioConfig_maxDataFields() * prec <= INT_MAX we are fine. Does that always hold?

If so, this check (together with (MP_DIGIT_MAX > ULLONG_MAX)) will prevent users from trying to aggregate to many bits/ints.

Reply:
> As long as "PrioConfig_maxDataFields() * prec <= INT_MAX" we are fine. Does that always hold?

I don't think that the above relation always holds, but I think that `PrioConfig_numIntEntries() * prec <= INT_MAX` probably does hold.

Either way, it might be good to add an assertion just to make sure that there isn't an integer overflow. Alternatively, we could put a global cap on `prec` everywhere (to something like 32, 64, or 128) just to be very sure that we will never even get close to an overflow on a 32- or 64-bit platform. Either way works for me.

Reply:
BBIT_MAX_PREC will usually be 63 now, since this is what fits in longs on most platforms. We could use unsigned long, but then I will need to cover different edge cases in the encoding, since the compiler optimizes some checks away. We could also use long long?

Reply:
Max'ing out at 63 bits sounds fine to me. That should handle all of the use cases that we have discussed.

Reply:
Missed that the results can hold at most an MP_DIGIT which is at most ULONG_MAX. I set BBIT_PREC_MAX down to 42, so we will be able to aggregate (2^64-1)/(2^42-1) =~ 4194304 submissions.

Reply:
Do you think thats a good choice here? If so, we can resolve this too.

Reply:
Yes, I think that's fine.